pkg/envtest/server.go

```diff
@@ -62,14 +62,17 @@ const (
 	defaultKubebuilderControlPlaneStopTimeout  = 20 * time.Second
 )
 
-// Default binary path for test framework
-func defaultAssetPath(binary string) string {
-	assetPath := os.Getenv(envKubebuilderPath)
-	if assetPath == "" {
-		assetPath = defaultKubebuilderPath
+// getBinAssetPath will return the path for the binary informed
+func (te *Environment) getBinAssetPath(binary string) string {
+	if hasBinary(binary, te.BinaryDirectoryPath) {
+		return filepath.Join(te.BinaryDirectoryPath, binary)
 	}
-	return filepath.Join(assetPath, binary)
 
+	valueFromEnvVar := os.Getenv(envKubebuilderPath)
+	if hasBinary(binary, valueFromEnvVar) {
+		return filepath.Join(valueFromEnvVar, binary)
+	}
```

Comment:
Typically, environment variables override hard-coded configuration, so this should be
```
func (te *Environment) getBinAssetPath(binary string) string {
	if valueFromEnvVar := os.Getenv(envKubebuilderPath); valueFromEnvVar != "" {
		return filepath.Join(valueFromEnvVar, binary)
	}
	if te.BinaryDirectoryPath != "" {
		return filepath.Join(te.BinaryDirectoryPath, binary)
	}
	return filepath.Join(defaultKubebuilderPath, binary)
}

Reply:
Interestingly the k8s integration package comment says the order is [reversed](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.7.0-alpha.6/pkg/internal/testing/integration/doc.go#L51-L66), which I find strange because environment variables are typically used as overrides. Any thoughts on this?

Reply:
The priority here is: 

0. the value set via the structure in this case `Environment{}` 
1. the specific envvars for each binary: `TEST_ASSET_KUBE_APISERVER`, `TEST_ASSET_ETCD`, `TEST_ASSET_KUBECTL`
2. the path for all env var: `KUBEBUILDER_ASSETS `
3. the default value which is currently `/usr/local/kubebuilder/bin`

which matches with other cases scenarios in place as you linked. So, because of this, I do not agree with the suggestion on https://github.com/kubernetes-sigs/controller-runtime/pull/1214#discussion_r520916615